exporter/kafkaexporter/pdata_marshaler.go

```diff
@@ -42,36 +42,79 @@ func newPdataLogsMarshaler(marshaler plog.Marshaler, encoding string) LogsMarsha
 	}
 }
 
+// KeyableMetricsMarshaler is an extension of the MetricsMarshaler interface intended to provide partition key capabilities
+// for metrics messages
+type KeyableMetricsMarshaler interface {
+	MetricsMarshaler
+	Key(attributes []string)
+}
+
 type pdataMetricsMarshaler struct {
-	marshaler pmetric.Marshaler
-	encoding  string
+	marshaler     pmetric.Marshaler
+	encoding      string
+	keyed         bool
+	keyAttributes []string
+}
+
+// Key configures the pdataMetricsMarshaler to set the message key on the kafka messages
+func (p *pdataMetricsMarshaler) Key(attributes []string) {
+	p.keyed = true
+	p.keyAttributes = attributes
 }
 
 func (p pdataMetricsMarshaler) Marshal(ld pmetric.Metrics, topic string) ([]*sarama.ProducerMessage, error) {
-	bts, err := p.marshaler.MarshalMetrics(ld)
-	if err != nil {
-		return nil, err
-	}
-	return []*sarama.ProducerMessage{
-		{
+	var msgs []*sarama.ProducerMessage
+	if p.keyed {
+		metrics := ld.ResourceMetrics()
+
+		for i := 0; i < metrics.Len(); i++ {
+			resourceMetrics := metrics.At(i)
+			var hash [16]byte
+			if len(p.keyAttributes) > 0 {
+				hash = pdatautil.MapHashSelectedKeys(resourceMetrics.Resource().Attributes(), p.keyAttributes)
+			} else {
+				hash = pdatautil.MapHash(resourceMetrics.Resource().Attributes())
+			}
+
+			newMetrics := pmetric.NewMetrics()
+			resourceMetrics.MoveTo(newMetrics.ResourceMetrics().AppendEmpty())
+
+			bts, err := p.marshaler.MarshalMetrics(newMetrics)
+			if err != nil {
+				return nil, err
+			}
+			msgs = append(msgs, &sarama.ProducerMessage{
+				Topic: topic,
+				Value: sarama.ByteEncoder(bts),
+				Key:   sarama.ByteEncoder(hash[:]),
```

Comment:
What's the purpose of using `[:]` instead of just passing in `hash`?

Reply:
hash has a signature of `[16]byte` but ByteEncoder is a `[]byte`